examples/scene/instanced_mesh.py

```diff
@@ -0,0 +1,116 @@
+# -*- coding: utf-8 -*-
+# -----------------------------------------------------------------------------
+# Copyright (c) Vispy Development Team. All Rights Reserved.
+# Distributed under the (new) BSD License. See LICENSE.txt for more info.
+# -----------------------------------------------------------------------------
+"""
+Instanced rendering of arbitrarily transformed meshes
+=====================================================
+"""
+
+from vispy import app, gloo, visuals, scene, use
+import numpy as np
+from scipy.spatial.transform import Rotation
+from vispy.io import read_mesh, load_data_file
+
+# full gl+ context is required for instanced rendering
+use(gl='gl+')
+
+
+vertex_shader = """
+// these attributes will be defined on an instance basis
+attribute vec3 shift;
+attribute vec4 color;
+attribute vec3 transform_x;
+attribute vec3 transform_y;
+attribute vec3 transform_z;
+
+varying vec4 v_color;
+
+void main() {
+    v_color = color;
+    // transform is generated from column vectors (new basis vectors)
+    // https://en.wikibooks.org/wiki/GLSL_Programming/Vector_and_Matrix_Operations#Constructors
+    mat3 instance_transform = mat3(transform_x, transform_y, transform_z);
+    vec3 pos_rotated = instance_transform * $position;
+    vec4 pos_shifted = vec4(pos_rotated + shift, 1);
+    gl_Position = $transform(pos_shifted);
+}
+"""
+
+fragment_shader = """
+varying vec4 v_color;
+
+void main() {
+  gl_FragColor = v_color;
+}
+"""
+
+
+class InstancedMeshVisual(visuals.Visual):
+    def __init__(self, vertices, faces, positions, colors, transforms, subdivisions=5):
+        visuals.Visual.__init__(self, vertex_shader, fragment_shader)
+
+        self.set_gl_state('translucent', depth_test=True, cull_face=True)
+        self._draw_mode = 'triangles'
+
+        # set up vertex and index buffer
+        self.vbo = gloo.VertexBuffer(vertices.astype(np.float32))
+        self.shared_program.vert['position'] = self.vbo
+        self._index_buffer = gloo.IndexBuffer(data=faces.astype(np.uint32))
+
+        # create a vertex buffer with a divisor argument of 1. This means that the
+        # attribute value is set to the next element of the array every 1 instance.
+        # The length of the array multiplied by the divisor determines the number
+        # of instances
+        self.shifts = gloo.VertexBuffer(positions.astype(np.float32), divisor=1)
+        self.shared_program['shift'] = self.shifts
+
+        # vispy does not handle matrix attributes (likely requires some big changes in GLIR)
+        # so we decompose it into three vec3; (column vectors of the matrix)
+        transforms = transforms.astype(np.float32)
+        self.transforms_x = gloo.VertexBuffer(transforms[..., 0], divisor=1)
+        self.transforms_y = gloo.VertexBuffer(transforms[..., 1], divisor=1)
+        self.transforms_z = gloo.VertexBuffer(transforms[..., 2], divisor=1)
```

Comment:
My guess is this is the problem regarding the dis-contiguous data. By doing the indexing there you're now causing striding by 3 (for the 3 elements X/Y/Z) and the VertexBuffer underneath is complaining that the data isn't contiguous and has to be copied.

https://github.com/vispy/vispy/blob/ded8eb9d7483255789dcac87b01399bd3925905b/vispy/gloo/buffer.py#L438-L442

Reply:
So I think if you did a `.copy()` here when you do the `[..., 0]` style indexing then you'd at least avoid the warning. If you could find a way to structure the array as `(3, M, N)` then do indexing `[0]` then you'd also avoid the copy.